src/de_impl_opt.rs

```diff
@@ -0,0 +1,442 @@
+use std::{f64, fmt};
+
+use crate::de::{self, Deserializer};
+
+/// De-serialize either a `str`, `i64`, `f64`, or `u64`
+/// as a *signed* value wrapped in [`Some`],
+/// and a `bool` or `null` value as [`None`].
+///
+/// # Returns
+/// A [`Some`] with the signed (`i64`) value of a string
+/// or number.
+///
+/// A [`None`] in the case of:
+///   * a `bool` value.
+///   * a `null` value.
+///   * any *de-serialization* errors.
+///     * ex. a string is non-empty and not a valid numeric value.
+///     * ex. the unsigned value `u64` *overflows* when converted to `i64`.
+///
+pub fn as_opt_i64<'de, D>(deserializer: D) -> Result<Option<i64>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    deserializer.deserialize_any(DeserializeOptionalI64WithVisitor)
+}
+
+/// De-serialize either a `str`, `u64`, `f64`, or `i64`
+/// as an *unsigned* value wrapped in [`Some`],
+/// and a `bool` or `null` value as [`None`].
+///
+/// # Returns
+/// A [`Some`] with the unsigned (`u64`) value of a string
+/// or number.
+///
+/// A [`None`] in the case of:
+///   * a `bool` value.
+///   * a `null` value.
+///   * any *de-serialization* errors.
+///     * ex. a string is non-empty and not a valid numeric value.
+///     * ex. the signed value `i64` represents a *negative* number.
+///
+pub fn as_opt_u64<'de, D>(deserializer: D) -> Result<Option<u64>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    deserializer.deserialize_any(DeserializeOptionalU64WithVisitor)
+}
+
+/// De-serialize either a `str`, `f64`, `u64`, or `i64`
+/// as a *float* value wrapped in [`Some`],
+/// and a `bool` or `null` value as [`None`].
+///
+/// # Returns
+/// A [`Some`] with the floating point (`f64`) value of a string
+/// or number.
+///
+/// A [`None`] in the case of:
+///   * a `bool` value.
+///   * a `null` value.
+///   * any *de-serialization* errors.
+///     * ex. a string is non-empty and not a valid numeric value.
+///
+pub fn as_opt_f64<'de, D>(deserializer: D) -> Result<Option<f64>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    deserializer.deserialize_any(DeserializeOptionalF64WithVisitor)
+}
+
+/// De-serialize either a `bool`, `str`, `u64`, or `f64`
+/// as a *boolean* value wrapped in [`Some`],
+/// and an `i64` or `null` value as [`None`].
+///
+/// # Truthy String Values
+/// > Note: the pattern matching is *case insensitive*, so `YES` or `yes`
+/// > works just the same.
+///
+/// These are the following "truthy" string values that result in a
```

Comment:
For completeness, also the `falsy` values need to be defined - since we're moving to three valued logic now. This seems to not have been required previously as everything that was not `truthy` was `falsy`. But now the formerly `falsy` need to be split into `falsy` and `None`

Reply:
Done @corneliusroemer ! Thanks for bringing this up. I have added a doc section for the function `as_opt_bool` documenting the "falsy" string values, as suggested.